gh-115999: Specialize `LOAD_ATTR` for instance and class receivers in free-threaded builds

mpage opened the pull request:
This PR finishes specialization for`LOAD_ATTR` in the free-threaded build by adding support for class and instance receivers.

The bulk of it is dedicated to making the specialized instructions and the specialization logic thread-safe. This consists of using atomics / locks in the appropriate places, avoiding races in specialization related to reloading versions, and ensuring that the objects stored in inline caches remain valid when accessed (by only storing deferred objects). See the section on "Thread Safety" below for more details.

Additionally, making this work required a few unrelated changes to fix existing bugs or work around differences between the two builds that results from only storing deferred values (which causes in specialization failures in the free-threaded build when a value that would be stored in the cache is not deferred):

- Fixed a bug in the cases generator where it wasn't treating macro tokens as terminators when searching for the assignment target of an expression involving `PyStackRef_FromPyObjectNew`.
- Refactored `test_descr.MiscTests.test_type_lookup_mro_reference` to work when specialization fails (and also be a behavorial test).  
- Temporarily skip `test_capi.test_type.TypeTests.test_freeze_meta` when running refleaks tests on free-threaded builds. Specialization failure triggers an existing bug.

## Single-threaded Performance

* [Performance](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241220-3.14.0a3+-d6d4c73-NOGIL#vs-base) is improved by ~12% on free-threaded builds.
* [Performance](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241220-3.14.0a3+-d6d4c73#vs-base) is ~neutral on default builds. (Technically the benchmark runner reports a 0.2% improvement, but that looks like noise to me.)

We're leaving a bit of perf on the table by only storing deferred objects: we can't specialize attribute lookups that resolve to class attributes (e.g. counters, settings). I haven't measured how much perf we're giving up, but I'd like to address that in a separate PR.

## Scalability

The `object_cfunction` and `pymethod` benchmarks are improved (1.4x slower -> 14.3x faster, 1.8x slower -> 13.0x faster, respectively). Other benchmarks appear unchanged.

I would expect that `cmodule_function` would also improve, but it looks like the benchmark is bottlenecked on increfing the `int.__floor__` method that is returned from the call to `_PyObject_LookupSpecial` in `math_floor` (the incref happens in `_PyType_LookupRef`, which is called by `PyObject_LookupSpecial`):

https://github.com/python/cpython/blob/3879ca0100942ae15a09ac22889cbe3e46d424eb/Modules/mathmodule.c#L1178

Raw numbers:

```
Running benchmarks with 28 threads
                   Merge-base    This PR
object_cfunction    1.4x slower  14.3x faster
cmodule_function    1.7x slower   1.7x slower
mult_constant      12.8x faster  13.6x faster
generator          11.5x faster  12.2x faster
pymethod            1.8x slower  13.4x faster
pyfunction         12.2x faster  13.0x faster
module_function    14.5x faster  14.8x faster
load_string_const  13.0x faster  13.2x faster
load_tuple_const   12.7x faster  14.1x faster
create_pyobject    12.8x faster  13.5x faster
create_closure     13.9x faster  14.8x faster
create_dict        10.5x faster  13.3x faster
thread_local_read   3.9x slower   4.0x slower
```

## Thread Safety

Thread safety of specialized instructions is addressed in a few different ways:

- Use atomics where needed.
- Lock the instance dictionary in `_LOAD_ATTR_WITH_HINT`.
- Only store deferred objects in the inline cache. All uops that retrieve objects from the cache are preceded by type version guards, which ensure that the (deferred) value retrieved from the cache is valid. The type version will be mutated before destroying the reference in the type. If the type held the last reference, then GC will need to run in order to reclaim the object. GC requires stopping the world, so if GC reclaims the object then we will see the new type version and the guard will fail.

Thread safety of specialization is addressed using similar techniques:

- Atomics / locking is used to access shared state.
- Specialization code is refactored to read versions at the start of specialization and store the same version in caches (as opposed to rereading the version). This ensures that the specialization is consistent with the version by avoiding races where the type or shared keys dictionary changes after we've queried it.

## Stats

Following the [instructions](https://github.com/python/cpython/blob/2a66dd33dfc0b845042da9bb54aaa4e890733f54/Python/specialize.c#L1516-L1518) in the comment preceding `specialize_attr_loadclassattr`, I collected stats for the default build for both this PR and its merge base using `./python -m test_typing test_re test_dis test_zlib` and compared them using `Tools/scripts/summarize_stats.py`.  The results for `LOAD_ATTR` are nearly identical and are consistent with results from comparing the merge base against itself:

- [Comparison](https://htmlpreview.github.io/?https://gist.githubusercontent.com/mpage/c19759d3ab374de3f6e17ccc2a85bd32/raw/472bc29bd5352402db39d542f2ffa4f3347d4c11/merge_base_vs_pr.html)
- [Raw stats for merge base](https://gist.github.com/mpage/e65379857b9f7b2822f30007d193cef8)
- [Raw stats for PR](https://gist.github.com/mpage/a636f0d7c973170c95f12ca3e7820df6)



<!-- gh-issue-number: gh-115999 -->
* Issue: gh-115999
<!-- /gh-issue-number -->

mpage committed:
Add _PyDictKeys_StringLookupAndVersion

Look up a unicode key in an all unicode keys object along with the
keys version, assigning one if not present.

We need a keys version that is consistent with presence of the key
for use in the guards.

mpage committed:
Pass shared keys version to specialization

Reading the shared keys version and looking up the key need to be performed
atomically. Otherwise, a key inserted after the lookup could race with reading
the version, causing us to incorrectly specialize that no key shadows the
descriptor.

mpage committed:
Add support for enabling each of the instance attribute kinds

Everything starts out disabled

mpage committed:
Only cache deferred descriptors

mpage committed:
Make analyze_descriptor thread-safe

* Check that the type hasn't changed across lookups
* Descr is now an owned ref

mpage committed:
Use an atomic load for GUARD_TYPE_VERSION

mpage committed:
Use atomics to load valid bit for inline values in _GUARD_DORV_VALUES_INST_ATTR_FROM_DICT

mpage committed:
Use atomic to load keys version in _GUARD_KEYS_VERSION

mpage committed:
Use an atomic load for managed dict in `_CHECK_ATTR_METHOD_LAZY_DICT`

mpage committed:
Enable specialization of method loads

mpage committed:
Use atomics for fetching type flags

mpage committed:
Get a strong reference to dict in instance_has_key

mpage committed:
Split specialize_dict_access

mpage committed:
Pass type version to specialize_dict_access

mpage committed:
Take a critical section around dict specialization

mpage committed:
Use thread-safe version of _PyDictKeys_StringLookup

mpage committed:
Use atomic load in _CHECK_MANAGED_OBJECT_HAS_VALUES

mpage committed:
Make _LOAD_ATTR_INSTANCE_VALUE thread-safe

- Use atomic load for value
- Use _Py_TryIncrefCompareStackRef for incref

mpage committed:
Make _LOAD_ATTR_WITH_HINT thread-safe

mpage committed:
Specialize instance accesses

mpage committed:
Specialize LOAD_ATTR_SLOT

- Use atomics and _Py_TryIncrefCompareStackRef in _LOAD_ATTR_SLOT
- Pass type version during specialization

mpage committed:
Enable LOAD_ATTR_PROPERTY

- Check that fget is deferred
- Pass tp_version

mpage committed:
Checkpoint LOAD_ATTR_GETATTRIBUTE_OVERRIDEN

mpage committed:
Lock dict in instance_has_key

mpage committed:
Lock dict instead of owner when specializing for dict access

mpage committed:
Use atomic load for valid bit

mpage committed:
Use _PyDictKeys_StringLookup and lock around it, rather than wasting out param

mpage committed:
Fix cases_generator bug

Macros should be treated as terminators when searching for the assignment
target of an expression involving PyStackRef_FromPyObjectNew

mpage committed:
Fix load

mpage committed:
Remove FT_UNIMPLEMENTED

All instance loads are complete!

mpage committed:
Specialize class attribute loads

- Use atomics where appropriate
- Only cache deferred values

mpage committed:
Fix test_type_lookup_mro_reference

If specialization fails, `X.mykey` will be performed again using the
new mro set by __eq__.

Check that the process doesn't crash, since that's what we care about.

mpage committed:
Remove TYPE_CHANGED

Always use the type version from the __getattribute__ lookup (even if
its zero). The guard will fail if the type changes afterwards.

mpage committed:
Misc clean ups

mpage committed:
Check that inline values still valid after taking CS

mpage committed:
Enable test_opcache tests

mpage committed:
Skip test that triggers gh-127773 in refleak tests

mpage committed:
Merge branch 'main' into gh-115999-load-attr

mpage committed:
Add test for LOAD_ATTR_CLASS_WITH_METACLASS_CHECK

mpage committed:
Undo workaround for cases generator bug

mpage committed:
Fix whitespace after merge

mpage committed:
Double check that dict hasn't changed after locking it

mpage committed:
Use correct type for loading type version

mpage committed:
Remove unnecessary comma

mpage committed:
Use atomics when loading oparg

mpage committed:
Fix formatting

mpage committed:
Always return type version from analyze_descriptor_load

This matches the previous implementation and causes failures to
specialize due to the presence of both __getattr__ and __getattribute__
to be classified correctly (rather than being classified as being out
of type versions).

colesbury commented on Python/bytecodes.c:
```diff
@@ -2222,16 +2227,36 @@ dummy_func(
+            }
             STAT_INC(LOAD_ATTR, hit);
             Py_INCREF(attr_o);
             attr = PyStackRef_FromPyObjectSteal(attr_o);
```
The `Py_INCREF` and `PyStackRef_FromPyObjectSteal` should be combined into `PyStackRef_FromPyObjectNew(attr_o)`

colesbury commented on Objects/dictobject.c:
```diff
@@ -1129,6 +1129,23 @@ dictkeys_generic_lookup(PyDictObject *mp, PyDictKeysObject* dk, PyObject *key, P
 }
 
+static Py_hash_t
+check_keys_and_hash(PyDictKeysObject *dk, PyObject *key) {
```
```suggestion
check_keys_and_hash(PyDictKeysObject *dk, PyObject *key)
{
```

colesbury commented on Python/bytecodes.c:
```diff
@@ -2222,16 +2227,36 @@ dummy_func(
             PyObject *attr_o;
 
             PyDictObject *dict = _PyObject_GetManagedDict(owner_o);
```
I think we should push the dict to the stack in `_CHECK_ATTR_WITH_HINT`:

* It will avoid a redundant load in the free threading build
* I thin the managed dict could be cleared (set to NULL) in between the two loads in the free threading build.

colesbury commented on Objects/dictobject.c:
```diff
@@ -1129,6 +1129,23 @@ dictkeys_generic_lookup(PyDictObject *mp, PyDictKeysObject* dk, PyObject *key, P
+        if (hash == -1) {
+            PyErr_Clear();
+            return -1;
+        }
```
`PyUnicode_Type.tp_hash` can never fail. I think `assert(hash != -1)` would be sufficient

colesbury commented on Python/bytecodes.c:
```diff
@@ -2222,16 +2227,36 @@ dummy_func(
+            }
             PyObject *name = GETITEM(FRAME_CO_NAMES, oparg>>1);
-            DEOPT_IF(!DK_IS_UNICODE(dict->ma_keys));
+            if (!DK_IS_UNICODE(dict->ma_keys)) {
```
I think this should be an explicit check that the kind is `DICT_KEYS_UNICODE`. Currently, this allows for split keys, which will deopt later when the `me_value` is `NULL`. Excluding split keys early makes the thread-safety analysis easier (and isn't any more expensive).

colesbury commented on Python/bytecodes.c:
```diff
@@ -2222,16 +2227,36 @@ dummy_func(
 
             PyDictObject *dict = _PyObject_GetManagedDict(owner_o);
-            DEOPT_IF(hint >= (size_t)dict->ma_keys->dk_nentries);
+            DEOPT_IF(!LOCK_OBJECT(dict));
```
I'm not thrilled with the lock here -- we've tried to avoid locking in the fast path dictionary accesses -- but given the overall perf gains, we can come back to this later.

colesbury commented on Objects/dictobject.c:
```diff
@@ -1167,19 +1184,27 @@ unicodekeys_lookup_split(PyDictKeysObject* dk, PyObject *key, Py_hash_t hash)
+}
+
+Py_ssize_t
+_PyDictKeys_StringLookupAndVersion(PyDictKeysObject* dk, PyObject *key, uint32_t *version)
```
```suggestion
_PyDictKeys_StringLookupAndVersion(PyDictKeysObject *dk, PyObject *key, uint32_t *version)
```

colesbury reviewed (commented):
Overall looks great!

I think I unintentionally changed the `object_cfunction` when cleaning it up for inclusion in CPython. It previously had a `* 1.0` to coerce the value to a float and avoid the `_PyObject_LookupSpecial`.

(We should deal with the `_PyObject_LookupSpecial` bottlenecks, but it's a lower priority)

mpage committed:
Merge branch 'main' into gh-115999-load-attr-instance-merged

mpage committed:
Combine incref/steal into new stackref

mpage committed:
Fix formatting

mpage committed:
Remove unnecessary error check for PyUnicode_Type.tp_hash

mpage committed:
Check keys kind explicitly

mpage committed:
Pass dict from `_CHECK_ATTR_WITH_HINT` to `_LOAD_ATTR_WITH_HINT`

mpage committed:
Fix unused variable warning

colesbury commented on Python/specialize.c:
```diff
@@ -932,16 +935,20 @@ analyze_descriptor_load(PyTypeObject *type, PyObject *name, PyObject **descr) {
+        if (descriptor == _PyType_Lookup(&PyBaseObject_Type, name)) {
+            return DUNDER_CLASS;
+        }
     }
```
Is this different from the previous call to `descriptor_is_class()`? What am I missing?

mpage replied:
You're not missing anything - the refactor to use `descriptor_is_class` was lost while resolving merge conflicts. Fixed.

colesbury commented on Python/specialize.c:
```diff
@@ -1074,44 +1089,47 @@ static int specialize_class_load_attr(PyObject* owner, _Py_CODEUNIT* instr, PyOb
         return false;
     }
-    if (dict->ma_values) {
+    if (FT_ATOMIC_LOAD_PTR(dict->ma_values)) {
```
I think the thread-safety would be easier to reason about if this check was pushed inside the critical section.

colesbury commented on Python/specialize.c:
```diff
@@ -1517,29 +1573,41 @@ specialize_class_load_attr(PyObject *owner, _Py_CODEUNIT *instr,
+    unsigned long tp_flags = PyType_GetFlags(owner_cls);
+    if (tp_flags & Py_TPFLAGS_INLINE_VALUES) {
+        #ifndef Py_GIL_DISABLED
         PyDictKeysObject *keys = ((PyHeapTypeObject *)owner_cls)->ht_cached_keys;
```
unused variable warning on `keys` (presumably when assertions are disabled)

nascheme commented on Lib/test/test_opcache.py:
```diff
@@ -714,11 +714,13 @@ def write(items):
-                a = object()
+                # a must be set to an instance that uses deferred reference
+                # counting in free-threaded builds
+                a = type("Foo", (object,), {})
```
This looks a bit mysterious to me.  Does this do something different than what `class a: [...]` does?  Calling it an instance (which techically true) is a bit confusing too.  I would just say a "set to an object that uses ...".   It's a class object.  If you need it written like this, I think a little helper function with a docstring explaining why it must be done that way would help:
```
def make_deferred_ref_count_obj():
    return type("Foo", (object,), {})
 ```

nascheme replied:
Ah, I see later there is `item.a = type("Foo", (object,), {})` so you want an expression not a statement.

mpage committed:
Update number of specialization failure kinds

mpage committed:
Clarify construction of deferred object

mpage committed:
Add suppression for _PyUnicode_CheckConsistency

mpage committed:
Merge branch 'main' into gh-115999-load-attr-instance-merged

mpage committed:
Fix compiler warning

mpage committed:
Make it easier to reason about thread-safety of instance_has_key

mpage committed:
Restore refactor lost in merge

Use descriptor_is_class

colesbury commented on Tools/tsan/suppressions_free_threading.txt:
```diff
@@ -43,6 +43,8 @@ race_top:write_thread_id
 # Only seen on macOS, sample: https://gist.github.com/aisk/dda53f5d494a4556c35dde1fce03259c
 race_top:set_default_allocator_unlocked
+# gh-128212
+race_top:_PyUnicode_CheckConsistency
```
This should be fixed now if you merge main into your PR again.

mpage committed:
Merge branch 'main' into gh-115999-load-attr-instance-merged

mpage committed:
Revert "Add suppression for _PyUnicode_CheckConsistency"

This reverts commit 6f8aebf58e1302b9ae740d45b720ec5168ee30c7.

colesbury reviewed (approved):
Nice! LGTM!

markshannon commented on Objects/dictobject.c:
```diff
@@ -1129,6 +1129,21 @@ dictkeys_generic_lookup(PyDictObject *mp, PyDictKeysObject* dk, PyObject *key, P
 }
 
+static Py_hash_t
+check_keys_and_hash(PyDictKeysObject *dk, PyObject *key)
```
I can't tell from the name what this is checking.
Maybe `check_dict_is_unicode_only_and_key_has_hash_defined`?
Which leads me to the question: why check both of these things in a single function. The two checks appear unrelated.

mpage replied:
> Which leads me to the question: why check both of these things in a single function.

This is factoring out some common code into a helper. I'll split it into two helpers with more descriptive names.

markshannon commented on Python/bytecodes.c:
```diff
@@ -2147,18 +2147,23 @@ dummy_func(
             PyObject *owner_o = PyStackRef_AsPyObjectBorrow(owner);
             PyObject **value_ptr = (PyObject**)(((char *)owner_o) + offset);
-            PyObject *attr_o = *value_ptr;
+            PyObject *attr_o = FT_ATOMIC_LOAD_PTR_ACQUIRE(*value_ptr);
```
How do we know that the values array is still valid at this point?

colesbury replied:
There's a case that we are not handling correctly here: if, after loading `value_ptr`, another thread invalidates the inline values, modifies an attribute, and reallocates another object at the same memory address as the previous attribute, we can incorrectly return a value that wasn't the attribute at that offset.

We need to either:

* Re-check `_PyObject_InlineValues(owner_o)->valid` at the end of this handler and clean-up and deopt if it's not still true.
* Set the contents of the inlines values array to NULL when we mark it as invalid so that the `_Py_TryIncrefCompareStackRef` handles this case and correctly deopts.

I think the second options is probably simpler to implement and more efficient given that `_LOAD_ATTR_INSTANCE_VALUE` is common and invalidating inline values is relatively rare.

markshannon replied:
Invalidating inline values might not be that rare.
I suggest gathering some stats before deciding how to handle this.

colesbury replied:
From 
https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20250104-3.14.0a3%2B-0cafa97/bm-20250104-azure-x86_64-python-0cafa97932c6574734bb-3.14.0a3%2B-0cafa97-pystats.md:

LOAD_ATTR_INSTANCE_VALUE:  4,922,474,147
Materialize dict (on request) | 4,444,396 | 2.4%
Materialize dict (new key) | 476,375 | 0.3%
Materialize dict (too big) | 4,884 | 0.0%
Materialize dict (str subclass) | 0 | 0.0%

So, LOAD_ATTR_INSTANCE_VALUE is somewhere between 1,000x-10,000x more frequent than invalidating inline values. (Invalidating inline values requires first materializing the dictionary, but materializing the dictionary doesn't always invalidate the inline values).

markshannon commented on Python/bytecodes.c:
```diff
@@ -2147,18 +2147,23 @@ dummy_func(
+            PyObject *attr_o = FT_ATOMIC_LOAD_PTR_ACQUIRE(*value_ptr);
             DEOPT_IF(attr_o == NULL);
+            #ifdef Py_GIL_DISABLED
+            if (!_Py_TryIncrefCompareStackRef(value_ptr, attr_o, &attr)) {
```
What does `_Py_TryIncrefCompareStackRef` do?
Function names need to be self explanatory. The result of this function does not depend on a comparison, but on whether something has been modified. Maybe `_Py_TryIncrefIfPointersConsistentStackRef` as it only works if the `PyObject ** and PyObject *pointers are consistent.

I know that `_Py_TryIncrefCompareStackRef` was not introduced in this PR, but it is critical to the understanding of this code.

colesbury replied:
It is an atomic compare-and-incref function, similar to how `_Py_atomic_compare_exchange` is an atomic compare-and-swap function. If the comparison succeeds, the value is incref'd and `1` is returned. `0` is returned on failure. 

"Compare" is a more appropriate term than "consistent" here.

markshannon replied:
I still think "Compare" is misleading. The name should say what the function does, not how it does it.

markshannon commented on Python/bytecodes.c:
```diff
@@ -2208,32 +2213,58 @@ dummy_func(
             PyObject *attr_o;
+            if (!LOCK_OBJECT(dict)) {
+                DEAD(dict);
+                POP_DEAD_INPUTS();
```
This seems rather obscure. I see why you need to pop `dict`, but `DEAD(dict); POP_DEAD_INPUTS()` seems a rather convoluted way to do this.

Since you adding a new code generator macro, why not `POP(dict)`?
I think that most people would expect that an explicitly popped value is no longer live, so there should be no need for a kill and a pop.

mpage replied:
Sure, I can look at that. `POP` is already taken by a macro in `ceval_macros.h`, but we could use `POP_INPUT`.

markshannon replied:
Aside:
We shouldn't be using `POP`, as the code generator should handle pops. We can remove `POP` in another PR.

markshannon reviewed (commented):
A lot of effort in this PR seems to go into `LOAD_ATTR_WITH_HINT`. There is a lot of locking and unlocking and extra machinery to support having a value on the stack between uops.
Is it worth it?

Generally, I think any effort spent on `LOAD_ATTR_WITH_HINT` is better spent on increasing the fraction of objects that use inlined values and can be handled by `LOAD_ATTR_INSTANCE_VALUE`.
It is not just that `LOAD_ATTR_INSTANCE_VALUE` is faster, it also means that the object is created faster and uses less memory.

mpage committed:
Merge branch 'main' into gh-115999-load-attr-instance-merged

mpage committed:
Merge branch 'main' into gh-115999-load-attr-instance-merged

mpage committed:
Split check_keys_and_hash

These are two logically separate operations

mpage committed:
Clear inline values upon invalidation

This avoids a race where another thread invalidates the values, overwrites
an attribute stored in the values, and allocates a new object at the address
present in the values.

mpage committed:
Replace kill/pop_dead_inputs with pop_input

mpage committed:
Fix formatting

mpage committed:
Fix visibility

colesbury commented on Python/bytecodes.c:
```diff
@@ -2216,9 +2221,8 @@ dummy_func(
-            DEAD(mod_keys);
             // Clear mod_keys from stack in case we need to deopt
-            POP_DEAD_INPUTS();
+            POP_INPUT();
```
I think these macros are a bit easier to read when they include the name of the relevant argument, e.g., `POP_INPUT(mod_keys)`, and the code generator ensures that the name matches the top of stack.

colesbury commented on Objects/dictobject.c:
```diff
@@ -1951,6 +1951,17 @@ build_indices_unicode(PyDictKeysObject *keys, PyDictUnicodeEntry *ep, Py_ssize_t
+    assert(values->embedded);
+    FT_ATOMIC_STORE_UINT8(values->valid, 0);
+    for (int i = 0; i < values->capacity; i++) {
+        FT_ATOMIC_STORE_PTR(values->values[i], NULL);
```
This can be `FT_ATOMIC_STORE_PTR_RELEASE`.

I think the write to `valid` can be weakened too, but I'm less sure about the ordering and it's only one write (instead of up to 30) so I'm not going to worry about it.

mpage committed:
Merge branch 'main' into gh-115999-load-attr-instance-merged

mpage committed:
Use release instead of seq_cst

mpage committed:
Make POP_INPUT take the name of the TOS

mpage committed:
Merge branch 'main' into gh-115999-load-attr-instance-merged

mpage committed:
Use DECREF_INPUTS() instead of manually closing inputs
